File: third-party/thrift/src/thrift/lib/cpp2/type/TypeRegistry.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdexcept>

#include <folly/Singleton.h>
#include <folly/lang/Exception.h>
#include <thrift/lib/cpp2/type/TypeRegistry.h>

namespace apache::thrift::type {
namespace detail {

FOLLY_EXPORT TypeRegistry& getGeneratedTypeRegistry() {
  struct GeneratedTag {};
  return folly::detail::createGlobal<TypeRegistry, GeneratedTag>();
}

} // namespace detail

AnyData TypeRegistry::store(ConstRef value, const Protocol& protocol) const {
  return storeImpl(value, protocol);
}

void TypeRegistry::load(const AnyData& data, Ref out) const {
  if (data.type() == Type::get<type::void_t>()) {
    if (out.type() != Type::get<type::void_t>()) {
      folly::throw_exception<std::bad_any_cast>();
    }
    return;
  }
  folly::io::Cursor cursor{&data.data()};
  getEntry(data.type()).getSerializer(data.protocol()).decode(cursor, out);
}

void TypeRegistry::load(const AnyData& data, AnyRef out) const {
  if (data.type() == Type::get<type::void_t>()) {
    if (out.type() != Type::get<type::void_t>()) {
      folly::throw_exception<std::bad_any_cast>();
    }
    return;
  }
  folly::io::Cursor cursor{&data.data()};
  getEntry(data.type()).getSerializer(data.protocol()).decode(cursor, out);
}

AnyValue TypeRegistry::load(const AnyData& data) const {
  if (data.type() == Type::get<type::void_t>()) {
    return {};
  }
  folly::io::Cursor cursor{&data.data()};
  return getEntry(data.type())
      .getSerializer(data.protocol())
      .decode(data.type(), cursor);
}

bool TypeRegistry::registerSerializer(
    const op::Serializer& serializer, const Type& type) {
  if (!type.isFull()) {
    folly::throw_exception<std::runtime_error>(
        "Can only register types with fill uris.");
  }
  return types_[type]
      .protocols.emplace(serializer.getProtocol(), &serializer)
      .second;
}

bool TypeRegistry::isRegistered(const Type& type) const {
  if (!type.isFull()) { // TODO: Implement look up by type hash.
    folly::throw_exception<std::out_of_range>("Not Implemented");
  }

  // Lookup by exact type.
  return types_.find(type) != types_.end();
}

auto TypeRegistry::getEntry(const Type& type) const -> const TypeEntry& {
  if (!type.isFull()) { // TODO(afuller): Implement look by type hash.
    folly::throw_exception<std::out_of_range>("Not Implemented");
  }

  // Lookup by exact type.
  auto itr = types_.find(type);
  if (itr != types_.end()) {
    return itr->second;
  }

  // TODO(afuller): Improve error message.
  folly::throw_exception<std::out_of_range>("Type not registered.");
}

const op::Serializer& TypeRegistry::TypeEntry::getSerializer(
    const Protocol& protocol) const {
  auto itr = protocols.find(protocol);
  if (itr == protocols.end()) {
    folly::throw_exception<std::out_of_range>(
        fmt::format("Protocol not registered {}.", protocol.name()));
  }
  return *itr->second;
}

} // namespace apache::thrift::type

File: third-party/thrift/src/thrift/lib/go/thrift/types/type_registry.go
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package types

import (
	"crypto/sha256"
	"strings"
	"sync"
)

var (
	typeRegistryMu     sync.RWMutex
	typeRegistry       = make(map[string]*TypeSpec)
	typeHash32Registry = make(map[[32]byte]string)
	typeHash16Registry = make(map[[16]byte]string)
	typeHash8Registry  = make(map[[8]byte]string)
)

// THIS FUNCTION IS FOR INTERNAL USE ONLY.
// InternalRegisterType registers a URI + type spec pair for Thrit Any support.
// This is intended to be called from generated code's init() function.
func InternalRegisterType(uri string, spec *TypeSpec) {
	const thriftURIPrefix = "fbthrift://"

	uri, _ = strings.CutPrefix(uri, thriftURIPrefix)
	h := sha256.New()
	h.Write([]byte(thriftURIPrefix + uri))
	hash := h.Sum(nil)
	hash32 := *(*[32]byte)(hash[0:32])
	hash16 := *(*[16]byte)(hash[0:16])
	hash8 := *(*[8]byte)(hash[0:8])

	typeRegistryMu.Lock()
	defer typeRegistryMu.Unlock()

	typeRegistry[uri] = spec
	typeHash32Registry[hash32] = uri
	typeHash16Registry[hash16] = uri
	typeHash8Registry[hash8] = uri
}

// THIS FUNCTION IS FOR INTERNAL USE ONLY.
// InternalTypeRegistryGetFromURI returns the type spec for a given URI, if it exists.
func InternalTypeRegistryGetFromURI(uri string) (*TypeSpec, bool) {
	typeRegistryMu.RLock()
	defer typeRegistryMu.RUnlock()

	res, ok := typeRegistry[uri]
	return res, ok
}

// THIS FUNCTION IS FOR INTERNAL USE ONLY.
// InternalTypeRegistryGetFromHash returns the type spec for a given hash, if it exists.
func InternalTypeRegistryGetFromHash(hash []byte) (*TypeSpec, bool) {
	typeRegistryMu.RLock()
	defer typeRegistryMu.RUnlock()

	var uri string
	var ok bool
	switch len(hash) {
	case 32:
		hash32 := *(*[32]byte)(hash[0:32])
		uri, ok = typeHash32Registry[hash32]
	case 16:
		hash16 := *(*[16]byte)(hash[0:16])
		uri, ok = typeHash16Registry[hash16]
	case 8:
		hash8 := *(*[8]byte)(hash[0:8])
		uri, ok = typeHash8Registry[hash8]
	}

	if !ok {
		return nil, false
	}

	res, ok := typeRegistry[uri]
	return res, ok
}
